authority/provisioner/nebula.go

```diff
@@ -0,0 +1,442 @@
+	// For example {{ .AuthorizationCrt.Details.Groups }} can be used to get all the groups.
+	data.SetAuthorizationCertificate(crt)
+
+	templateOptions, err := TemplateSSHOptions(p.Options, data)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(signOptions,
+		templateOptions,
+		// Checks the validity bounds, and set the validity if has not been set.
+		&sshLimitDuration{p.claimer, crt.Details.NotAfter},
+		// Validate public key.
+		&sshDefaultPublicKeyValidator{},
+		// Validate the validity period.
+		&sshCertValidityValidator{p.claimer},
+		// Require all the fields in the SSH certificate
+		&sshCertDefaultValidator{},
+	), nil
+}
+
+// AuthorizeRenew returns an error if the renewal is disabled.
+func (p *Nebula) AuthorizeRenew(ctx context.Context, crt *x509.Certificate) error {
+	if p.claimer.IsDisableRenewal() {
+		return errs.Unauthorized("renew is disabled for nebula provisioner '%s'", p.GetName())
+	}
+	return nil
+}
+
+// AuthorizeRevoke returns an error if the token is not valid.
+func (p *Nebula) AuthorizeRevoke(ctx context.Context, token string) error {
+	return p.validateToken(token, p.audiences.Revoke)
+}
+
+// AuthorizeSSHRevoke returns an error if SSH is disabled or the token is invalid.
+func (p *Nebula) AuthorizeSSHRevoke(ctx context.Context, token string) error {
+	if !p.claimer.IsSSHCAEnabled() {
+		return errs.Unauthorized("ssh is disabled for nebula provisioner '%s'", p.Name)
+	}
+	if _, _, err := p.authorizeToken(token, p.audiences.SSHRevoke); err != nil {
+		return err
+	}
+	return nil
+}
+
+// AuthorizeSSHRenew returns an unauthorized error.
+func (p *Nebula) AuthorizeSSHRenew(ctx context.Context, token string) (*ssh.Certificate, error) {
+	return nil, errs.Unauthorized("nebula provisioner does not support SSH renew")
+}
+
+// AuthorizeSSHRekey returns an unauthorized error.
+func (p *Nebula) AuthorizeSSHRekey(ctx context.Context, token string) (*ssh.Certificate, []SignOption, error) {
+	return nil, nil, errs.Unauthorized("nebula provisioner does not support SSH rekey")
+}
+
+func (p *Nebula) validateToken(token string, audiences []string) error {
+	_, _, err := p.authorizeToken(token, audiences)
+	return err
+}
+
+func (p *Nebula) authorizeToken(token string, audiences []string) (*cert.NebulaCertificate, *jwtPayload, error) {
+	jwt, err := jose.ParseSigned(token)
+	if err != nil {
+		return nil, nil, errs.UnauthorizedErr(err, errs.WithMessage("failed to parse token"))
+	}
+
+	// Extract nebula certificate
+	nbc, ok := jwt.Headers[0].ExtraHeaders[NebulaCertHeader]
+	if !ok {
+		return nil, nil, errs.Unauthorized("failed to parse token: nbc header is missing")
+	}
+	s, ok := nbc.(string)
+	if !ok {
+		return nil, nil, errs.Unauthorized("failed to parse token: nbc header is not valid")
+	}
+	b, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		return nil, nil, errs.UnauthorizedErr(err, errs.WithMessage("failed to parse token: nbc header is not valid"))
```

Comment:
This is because we haven't yet changed the interface of that method. We will change it to it will be like:
```
func Unauthorized(format string, args ...interface{}) error
func UnauthorizedErr(err error, format string, args ...interface{}) error
```

Other errors like BadRequest and Forbidden already work like that.